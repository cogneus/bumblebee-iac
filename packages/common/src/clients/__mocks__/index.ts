export * from '../db/__mocks__'
export * from '../secrets/__mocks__'
export * from '../ssm/__mocks__'
