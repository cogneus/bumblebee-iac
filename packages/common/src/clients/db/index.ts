export * from './types'
export * from './db-client.class'
export * from './make-update-expression.helper'
export * from './get-record-limit.helper'
