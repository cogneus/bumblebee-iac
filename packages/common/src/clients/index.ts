export * from './db'
export * from './secrets'
export * from './ssm'
export * from './ensure-client'
export * from './types'
export * from './__mocks__'
export * from './destroy-clients'
