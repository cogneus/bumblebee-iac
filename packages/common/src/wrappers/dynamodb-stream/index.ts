export * from './dynamodb-stream.wrapper'
export * from './types/dynamo-db-stream-wrapper.type'
