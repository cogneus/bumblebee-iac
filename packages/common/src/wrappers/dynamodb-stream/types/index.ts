export * from './dynamo-db-recordhandler.type'
export * from './dynamo-db-stream-dependencies.interface'
export * from './dynamo-db-stream-wrapper.type'
export * from './dynamo-db-stream-wrapper-params.interface'
