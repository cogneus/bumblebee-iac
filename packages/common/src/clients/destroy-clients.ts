import { Clients } from './types'

export const destroyClients = (clients: Clients) => {
  if (clients) {
    Object.values(clients).forEach((client) => client?.destroy())
  }
}
