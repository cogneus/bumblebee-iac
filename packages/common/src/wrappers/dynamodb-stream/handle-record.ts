import { unmarshall } from '@aws-sdk/util-dynamodb'
import { AttributeValue } from '@aws-sdk/client-dynamodb'
import { DynamoDBStreamHandlerParams } from './types'
import { createWrapperContext } from '../utils'
import { destroyClients } from '../../clients'

export const handleRecord = async <TItem>({
  record,
  serviceName,
  region,
  env,
  logName,
  logLevel,
  context,
  handler,
  getClientOptions,
}: DynamoDBStreamHandlerParams<TItem>) => {
  const serviceContext = createWrapperContext({
    context,
    serviceName,
    region,
    env,
    logName,
    logLevel,
    logInfo: { record },
    clientOptions: getClientOptions(record),
  })
  const { log } = serviceContext
  try {
    const { eventName, dynamodb: { NewImage } = {} } = record
    const item = NewImage
      ? (unmarshall(NewImage as { [key: string]: AttributeValue }) as TItem)
      : null
    await handler({
      serviceContext,
      item,
      eventName,
    })
    log.debug('Process complete')
  } catch (ex) {
    log.warn(`Process failed, ${ex.message}`)
    log.error(ex)
  } finally {
    destroyClients(serviceContext.clients)
  }
}
