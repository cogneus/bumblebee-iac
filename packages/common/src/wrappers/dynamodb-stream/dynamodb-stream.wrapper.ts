import 'source-map-support/register'
import { DynamoDBStreamWrapper, DynamoDBStreamWrapperParams } from './types'
import { ServiceContext } from '../../types'
import { handleRecord } from './handle-record'

export const dynamoDBStreamWrapper: DynamoDBStreamWrapper =
  <TItem>(params: DynamoDBStreamWrapperParams<TItem>) =>
  async (event, context: ServiceContext) => {
    if (event.Records?.length) {
      for (const record of event.Records) {
        await handleRecord({
          ...params,
          record,
          context,
        })
      }
    }
  }
