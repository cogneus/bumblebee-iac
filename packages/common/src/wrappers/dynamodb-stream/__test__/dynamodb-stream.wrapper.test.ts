import { DynamoDBRecord, DynamoDBStreamEvent } from 'aws-lambda'
import { dynamoDBStreamWrapper } from '../dynamodb-stream.wrapper'
import { ClientOptions } from '../../../clients'

const baseRecord: DynamoDBRecord = {
  awsRegion: 'any',
  eventSource: '',
  eventSourceARN: 'a207919-vespa-ci-test-main-',
}

const baseStagingRecord: DynamoDBRecord = {
  ...baseRecord,
  eventSourceARN: 'a207919-vespa-ci-test-staging-',
}

const baseUnknownRecord: DynamoDBRecord = {
  ...baseRecord,
  eventSourceARN: 'a207919-vespa-ci-test-blue-',
}

const baseUndefinedRecord: any = {
  ...baseRecord,
  eventSourceARN: undefined,
}

const clientOptions: ClientOptions = {
  db: {
    maxAttempts: 3,
    region: 'eu-west-1',
    requestTimeout: 3000,
    tables: {
      test: {
        name: 'test-table-name',
        indices: {
          testIndex: { name: 'test-index-name' },
        },
        itemLimits: { defaultLimit: 20, maxLimit: 100 },
      },
    },
  },
}

const baseImage = {
  PK: {
    S: 'template#test-template',
  },
  SK: {
    S: 'version#0000000000',
  },
  created: {
    S: '2023-02-19T14:46:24.905Z',
  },
}

const events: DynamoDBStreamEvent[] = [
  {
    Records: [
      {
        ...baseRecord,
        eventName: 'INSERT',
        dynamodb: {
          NewImage: baseImage,
        },
      },
      {
        ...baseRecord,
        eventName: 'MODIFY',
        dynamodb: {
          NewImage: baseImage,
        },
      },
    ],
  },
  {
    Records: [
      {
        ...baseRecord,
        eventName: 'REMOVE',
        dynamodb: {
          NewImage: baseImage,
        },
      },
      {
        ...baseRecord,
        eventName: 'MODIFY',
        dynamodb: {},
      },
    ],
  },
  {
    Records: [
      {
        ...baseStagingRecord,
        eventName: 'MODIFY',
        dynamodb: {
          NewImage: baseImage,
        },
      },
      {
        ...baseStagingRecord,
        eventName: 'MODIFY',
        dynamodb: {},
      },
    ],
  },
]

const invalidEvents: DynamoDBStreamEvent[] = [
  {
    Records: [
      {
        ...baseRecord,
        eventName: 'INSERT',
        dynamodb: {
          NewImage: baseImage,
        },
      },
      {
        ...baseUnknownRecord,
        eventName: 'MODIFY',
        dynamodb: {
          NewImage: baseImage,
        },
      },
    ],
  },
  {
    Records: [
      {
        ...baseRecord,
        eventName: 'INSERT',
        dynamodb: {
          NewImage: baseImage,
        },
      },
      {
        ...baseUndefinedRecord,
        eventName: 'MODIFY',
        dynamodb: {
          NewImage: baseImage,
        },
      },
    ],
  },
]

describe('dynamo db stream wrapper tests', () => {
  describe('should handle response, returns', () => {
    it.each(events)('results matching snapshot', async (event) => {
      jest.resetAllMocks()
      const results: any[] = []
      const wrapper = dynamoDBStreamWrapper({
        handler: async ({ item, eventName }) => {
          results.push({
            eventName,
            item,
          })
        },
        region: 'eu-west-1',
        env: 'dev',
        logName: 'wrapper-test',
        serviceName: 'wrapper-test',
        getClientOptions: () => clientOptions,
      })

      const cb: any = null
      await wrapper(event, {} as any, cb)
      expect(results).toMatchSnapshot()
    })
  })

  describe('should throw Error, returns', () => {
    it.each(invalidEvents)('results matching snapshot', async (event) => {
      jest.resetAllMocks()
      const results: any[] = []
      const wrapper = dynamoDBStreamWrapper({
        handler: async ({ item, eventName }) => {
          results.push({
            eventName,
            item,
          })
        },
        region: 'eu-west-1',
        env: 'dev',
        logName: 'wrapper-test',
        serviceName: 'wrapper-test',
        getClientOptions: () => clientOptions,
      })

      const cb: any = null
      await wrapper(event, {} as any, cb)
      expect(results).toMatchSnapshot()
    })
  })
})
