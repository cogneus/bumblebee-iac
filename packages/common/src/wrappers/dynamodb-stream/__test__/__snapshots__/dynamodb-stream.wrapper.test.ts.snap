// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamo db stream wrapper tests should handle response, returns results matching snapshot 1`] = `
[
  {
    "eventName": "INSERT",
    "item": {
      "PK": "template#test-template",
      "SK": "version#0000000000",
      "created": "2023-02-19T14:46:24.905Z",
    },
  },
  {
    "eventName": "MODIFY",
    "item": {
      "PK": "template#test-template",
      "SK": "version#0000000000",
      "created": "2023-02-19T14:46:24.905Z",
    },
  },
]
`;

exports[`dynamo db stream wrapper tests should handle response, returns results matching snapshot 2`] = `
[
  {
    "eventName": "REMOVE",
    "item": {
      "PK": "template#test-template",
      "SK": "version#0000000000",
      "created": "2023-02-19T14:46:24.905Z",
    },
  },
  {
    "eventName": "MODIFY",
    "item": null,
  },
]
`;

exports[`dynamo db stream wrapper tests should handle response, returns results matching snapshot 3`] = `
[
  {
    "eventName": "MODIFY",
    "item": {
      "PK": "template#test-template",
      "SK": "version#0000000000",
      "created": "2023-02-19T14:46:24.905Z",
    },
  },
  {
    "eventName": "MODIFY",
    "item": null,
  },
]
`;

exports[`dynamo db stream wrapper tests should throw Error, returns results matching snapshot 1`] = `
[
  {
    "eventName": "INSERT",
    "item": {
      "PK": "template#test-template",
      "SK": "version#0000000000",
      "created": "2023-02-19T14:46:24.905Z",
    },
  },
  {
    "eventName": "MODIFY",
    "item": {
      "PK": "template#test-template",
      "SK": "version#0000000000",
      "created": "2023-02-19T14:46:24.905Z",
    },
  },
]
`;

exports[`dynamo db stream wrapper tests should throw Error, returns results matching snapshot 2`] = `
[
  {
    "eventName": "INSERT",
    "item": {
      "PK": "template#test-template",
      "SK": "version#0000000000",
      "created": "2023-02-19T14:46:24.905Z",
    },
  },
  {
    "eventName": "MODIFY",
    "item": {
      "PK": "template#test-template",
      "SK": "version#0000000000",
      "created": "2023-02-19T14:46:24.905Z",
    },
  },
]
`;
